test/utils.dart

```diff
@@ -66,6 +67,40 @@ class _Parse extends Matcher {
   }
 }
 
+/// A matcher that validates the body of a multipart request after finalization.
+/// The string "{{boundary}}" in [pattern] will be replaced by the boundary
+/// string for the request, and LF newlines will be replaced with CRLF.
+/// Indentation will be normalized.
+Matcher bodyMatches(String pattern) => new _BodyMatches(pattern);
+
+class _BodyMatches extends Matcher {
+  final String _pattern;
+
+  _BodyMatches(this._pattern);
+
+  bool matches(item, Map matchState) {
+    if (item is! http.Request) return false;
+
+    var future = collectBytes(item.read()).then((bodyBytes) {
```

Comment:
Why not just use `Request.readAsString()` here?

Reply:
Because of this check. The string length isn't necessarily the same as the contentLength at this point.

```dart
expect(item.contentLength, equals(bodyBytes.length));
```